File: easypermissions/src/test/java/pub/devrel/easypermissions/testhelper/FragmentController.java
package pub.devrel.easypermissions.testhelper;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.testing.FragmentScenario;

/**
 * Helper class to allow starting Fragments, similar to the old SupportFragmentController.
 */
public class FragmentController<T extends Fragment> {

    private FragmentScenario<T> scenario;

    public FragmentController(Class<T> clazz) {
        scenario = FragmentScenario.launch(clazz);
    }

    public synchronized T resume() {
        final CompletableFuture<T> fragmentFuture = new CompletableFuture<>();

        scenario.onFragment(new FragmentScenario.FragmentAction<T>() {
            @Override
            public void perform(@NonNull T fragment) {
                fragmentFuture.complete(fragment);
            }
        });

        try {
            return fragmentFuture.get();
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        } catch (ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    public void reset() {
        scenario.recreate();
    }

}

File: easypermissions/src/test/java/pub/devrel/easypermissions/controllers/FragmentController.kt
/*
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package pub.devrel.easypermissions.controllers

import androidx.fragment.app.Fragment
import androidx.fragment.app.testing.FragmentScenario
import pub.devrel.easypermissions.annotations.Mockable
import java.util.concurrent.CompletableFuture

/**
 * Controller class to allow starting Fragments, similar to the old SupportFragmentController.
 */
@Mockable
class FragmentController<T : Fragment>(clazz: Class<T>) {

    private val scenario: FragmentScenario<T> = FragmentScenario.launch(clazz)

    @Synchronized
    fun resume(): T {
        val fragmentFuture = CompletableFuture<T>()
        scenario.onFragment { fragmentFuture.complete(it) }
        return fragmentFuture.get()
    }
}
